src/music/midifile.cpp

```diff
@@ -450,8 +794,304 @@ void MidiFile::MoveFrom(MidiFile &other)
+	size_t tracksizepos = ftell(f) - 4;
+
+	/* Write blocks in sequence */
+	uint32 lasttime = 0;
+	size_t nexttempoindex = 0;
+	for (size_t bi = 0; bi < this->blocks.size(); bi++) {
+	redoblock:
+		DataBlock &block = this->blocks[bi];
+		TempoChange &nexttempo = this->tempos[nexttempoindex];
+
+		uint32 timediff = block.ticktime - lasttime;
+
+		/* Check if there is a tempo change before this block */
+		if (nexttempo.ticktime < block.ticktime) {
+			timediff = nexttempo.ticktime - lasttime;
+		}
+
+		/* Write delta time for block */
+		lasttime += timediff;
+		bool needtime = false;
+		WriteVariableLen(f, timediff);
+
+		/* Write tempo change if there is one */
+		if (nexttempo.ticktime <= block.ticktime) {
+			byte tempobuf[6] = { MIDIST_SMF_META, 0x51, 0x03, 0, 0, 0 };
+			tempobuf[3] = (nexttempo.tempo & 0x00FF0000) >> 16;
+			tempobuf[4] = (nexttempo.tempo & 0x0000FF00) >>  8;
+			tempobuf[5] = (nexttempo.tempo & 0x000000FF);
+			fwrite(tempobuf, 1, 6, f);
+			nexttempoindex++;
+			needtime = true;
+		}
+		/* If a tempo change occurred between two blocks, rather than
+		 * at start of this one, start over with delta time for the block. */
+		if (nexttempo.ticktime < block.ticktime) {
+			goto redoblock;
+		}
+
+		/* Write each block data command */
+		byte *dp = block.data.Begin();
+		while (dp < block.data.End()) {
+			/* Always zero delta time inside blocks */
+			if (needtime) {
+				fwrite(&(bb = 0), 1, 1, f);
+			}
+			needtime = true;
+
+			/* Check message type and write appropriate number of bytes */
+			switch (*dp & 0xF0) {
+			case MIDIST_NOTEOFF:
+			case MIDIST_NOTEON:
+			case MIDIST_POLYPRESS:
+			case MIDIST_CONTROLLER:
+			case MIDIST_PITCHBEND:
+				fwrite(dp, 1, 3, f);
+				dp += 3;
+				continue;
+			case MIDIST_PROGCHG:
+			case MIDIST_CHANPRESS:
+				fwrite(dp, 1, 2, f);
+				dp += 2;
+				continue;
+			}
+
+			/* Sysex needs to measure length and write that as well */
+			if (*dp == MIDIST_SYSEX) {
+				fwrite(dp, 1, 1, f);
+				dp++;
+				byte *sysexend = dp;
+				while (*sysexend++ != MIDIST_ENDSYSEX);
+				ptrdiff_t sysexlen = sysexend - dp;
+				WriteVariableLen(f, sysexlen);
+				fwrite(dp, 1, sysexend - dp, f);
+				dp = sysexend;
+				continue;
+			}
+
+			/* Fail for any other commands */
+			fclose(f);
+			return false;
+		}
+	}
+
+	/* End of track marker */
+	fwrite(&(bb = 0x00), 1, 1, f);
+	fwrite(&(bb = MIDIST_SMF_META), 1, 1, f);
+	fwrite(&(bb = 0x2F), 1, 1, f);
+	fwrite(&(bb = 0x00), 1, 1, f);
+
+	/* Fill out the RIFF block length */
+	size_t trackendpos = ftell(f);
+	fseek(f, tracksizepos, SEEK_SET);
+	uint32 tracksize = trackendpos - tracksizepos - 4;
+	tracksize = TO_BE32(tracksize);
+	fwrite(&tracksize, 4, 1, f);
+
+	fclose(f);
+	return true;
+}
+
+/**
+ * Get the name of a Standard MIDI File for a given song.
+ * For songs already in SMF format, just returns the original.
+ * Otherwise the song is converted and a temporary file written,
+ * and the temporary file's name returned.
```

Comment:
(I'm getting into utter pedantry mode at this point) No need for that many line breaks, there are no character limits here :) (Though keep it reasonable - ~150 chars?)